com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/dao/AbstractDao.java

```diff
@@ -0,0 +1,46 @@
+package com.sap.cloud.lm.sl.cf.core.dao;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.sap.cloud.lm.sl.cf.core.dto.persistence.DtoWithPrimaryKey;
+
+public abstract class AbstractDao<T, D extends DtoWithPrimaryKey<P>, P> {
+
+    public void add(T object) {
+        getDtoDao().add(toDto(object));
+    }
+
+    public T update(P primaryKey, T newObject) {
+        D newDto = getDtoDao().update(primaryKey, toDto(newObject));
+        return fromDto(newDto);
+    }
+
+    public void remove(P primaryKey) {
+        getDtoDao().remove(primaryKey);
+    }
+
+    public List<T> findAll() {
+        return getDtoDao().findAll()
+            .stream()
+            .map(this::fromDto)
+            .collect(Collectors.toList());
+    }
+
+    public T find(P primaryKey) {
+        D dto = getDtoDao().find(primaryKey);
+        return dto != null ? fromDto(dto) : null;
+    }
+
+    protected List<T> fromDtos(List<D> dtos) {
+        return dtos.stream()
+            .map(this::fromDto)
+            .collect(Collectors.toList());
+    }
+
+    protected abstract AbstractDtoDao<D, P> getDtoDao();
```

Comment:
Why this is a method and not a filed in the class?
In that way the Decorator will be fully implemented.